src/coreclr/jit/lowerarmarch.cpp

```diff
@@ -507,14 +506,55 @@ void Lowering::LowerCast(GenTree* tree)
 
     assert(!varTypeIsSmall(srcType));
 
-    if (tmpType != TYP_UNDEF)
+    // Transform:
+    //
+    //    *  CAST      long <- ulong <- uint                   *  AND      long
+    //    \--*  CAST      int <- small-uint <- int     =>      +--*  X        int
+    //       \--*  X         int                               \--*  CNS_INT  long  0xFF/0xFFFF
+    //
+    if (comp->opts.OptimizationEnabled() && op1->OperIs(GT_CAST) && !tree->isContained() && tree->TypeIs(TYP_LONG) &&
+        (dstType == TYP_ULONG))
     {
-        GenTree* tmp = comp->gtNewCastNode(tmpType, op1, tree->IsUnsigned(), tmpType);
-        tmp->gtFlags |= (tree->gtFlags & (GTF_OVERFLOW | GTF_EXCEPT));
+        GenTreeCast* innerCast    = op1->AsCast();
+        var_types    innerDstType = innerCast->CastToType();
+        GenTree*     castOp       = innerCast->CastOp();
+        UINT32       maskSize     = 0;
+
+        if (varTypeIsSmall(innerDstType) && varTypeIsUnsigned(innerDstType) && !castOp->isContained() &&
+            !innerCast->isContained() && castOp->TypeIs(TYP_INT))
```

Comment:
Do we not need to check `gtOverflow` on the inner cast?

> `castOp->TypeIs(TYP_INT)`

Is CQ the driving factor for excluding small type here too? Would be worth a comment I think.

(Side note: I really like `innerCast`, much better than `cast2` 👍)